
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Phylogenetics in R &#8212; Biodiversity and Conservation Biology</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Macroecology analyses" href="../practical_3/practical_3.html" />
    <link rel="prev" title="Basics in R" href="../practical_1/practical_1.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">Biodiversity and Conservation Biology</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../intro.html">
   Introduction
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../practical_1/practical_1.html">
   Basics in R
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Phylogenetics in R
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../practical_3/practical_3.html">
   Macroecology analyses
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../practical_4/practical_4.html">
   Setting conservation priorities
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../links.html">
   Useful links
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/markdowns/practical_2/practical_2.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction-and-resources">
   1. Introduction and resources
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#starting-a-new-r-session">
     Starting a new R session
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-refresher-of-phylogenetic-trees">
   2. A refresher of phylogenetic trees
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#tree-parameters">
     Tree parameters
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#informatic-representations-of-tree">
     Informatic representations of tree
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#text-based-formats">
     Text based formats
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#edge-table">
     Edge table
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#records-pointers">
     Records &amp; pointers
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#basic-tree-viewing-in-r">
   3. Basic tree viewing in
   <code class="docutils literal notranslate">
    <span class="pre">
     R
    </span>
   </code>
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reading-in-a-phylogeny-from-a-file">
     Reading in a phylogeny from a file
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#manipulating-phylogenetic-trees-in-r">
   4. Manipulating phylogenetic trees in
   <code class="docutils literal notranslate">
    <span class="pre">
     R
    </span>
   </code>
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#adding-trait-data-to-trees-in-r">
   5. Adding trait data to trees in
   <code class="docutils literal notranslate">
    <span class="pre">
     R
    </span>
   </code>
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <section id="phylogenetics-in-r">
<h1>Phylogenetics in R<a class="headerlink" href="#phylogenetics-in-r" title="Permalink to this headline">¶</a></h1>
<section id="introduction-and-resources">
<h2>1. Introduction and resources<a class="headerlink" href="#introduction-and-resources" title="Permalink to this headline">¶</a></h2>
<p>This practical introduces you to basic phylogenetic computing in <code class="docutils literal notranslate"><span class="pre">R</span></code>. We
will review importing phylogenetic trees as data files, displaying
phylogenetic trees visually, and some basic evolutionary computations
that can be conducted with phylogenetic trees. This practical will
deliver some of the important background for Coursework 1. Below you
will find some of the relevant resources required for this practical.</p>
<p>Parts (sections 2,3,4) of this practical are written by <a class="reference external" href="http://nhcooper123.github.io/">Natalie
Cooper</a>. The original can be found
<a class="reference external" href="https://github.com/nhcooper123/TeachingMaterials/blob/master/PhD_Museum/VisualisingPhylo.Rmd">here</a>.
Whereas, parts (section 5 &amp; 6) of this practical are written by Adam
Devenish &amp; Rob Barber (<a class="reference external" href="mailto:a&#46;devenish&#37;&#52;&#48;imperial&#46;ac&#46;uk">mailto:a<span>&#46;</span>devenish<span>&#64;</span>imperial<span>&#46;</span>ac<span>&#46;</span>uk</a>)
(<a class="reference external" href="mailto:r&#46;barber19&#37;&#52;&#48;imperial&#46;ac&#46;uk">mailto:r<span>&#46;</span>barber19<span>&#64;</span>imperial<span>&#46;</span>ac<span>&#46;</span>uk</a>).</p>
<section id="starting-a-new-r-session">
<h3>Starting a new R session<a class="headerlink" href="#starting-a-new-r-session" title="Permalink to this headline">¶</a></h3>
<p>To plot phylogenies (or use any specialized analysis) in R, you need one
or more additional packages from the basic R installation. For this
practical you will need to load the following packages:</p>
<p>• ape</p>
<p>• phytools</p>
<p>Because we’ve started a new R session we shouldn’t have any packages
loaded. We need to load packages the same way as last session:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load packages.</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">phytools</span><span class="p">)</span>
</pre></div>
</div>
<p>You can think of <code class="docutils literal notranslate"><span class="pre">install.packages</span></code> like installing an app from the App
Store on your smart phone - you only do this once - and <code class="docutils literal notranslate"><span class="pre">library</span></code> as
being like pushing the app button on your phone - you do this every time
you want to use the app.</p>
<p>It’s also useful to remove any objects from our working directory before
starting a new project. You shouldn’t need to do this if you’ve just
started <code class="docutils literal notranslate"><span class="pre">RStudio</span></code>, but if you’ve been working on something before you
want to clear your workspace:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># rm removes objects. ls() returns everything in your environment.</span>
<span class="nf">rm</span><span class="p">(</span><span class="n">list</span><span class="o">=</span><span class="nf">ls</span><span class="p">())</span>
</pre></div>
</div>
</section>
</section>
<section id="a-refresher-of-phylogenetic-trees">
<h2>2. A refresher of phylogenetic trees<a class="headerlink" href="#a-refresher-of-phylogenetic-trees" title="Permalink to this headline">¶</a></h2>
<p>This section will review some basic aspects of phylogenetic trees and
introduce how trees are handled at the level of software. Because you
are now interacting with phylogenetic trees for things like plotting, it
is also helpful to know some of the names for parts of phylogenetic
trees used in computer science.</p>
<section id="tree-parameters">
<h3>Tree parameters<a class="headerlink" href="#tree-parameters" title="Permalink to this headline">¶</a></h3>
<p>A phylogenetic tree is an ordered, multifurcating graph with labelled
<strong>tips</strong> (or <strong>leaves</strong>) (and sometimes labelled histories). It
represents the relative degrees of relationships of species (i.e. tips
or OTUs). The graph consists of a series of <strong>branches</strong> (or <strong>edges</strong>)
with join successively towards <strong>nodes</strong> (or <strong>vertices</strong>, <em>sing.</em>
<strong>vertex</strong>). Each node is subtended by a single branch, representing the
lineage of ancestors leading to a node. The node is thus the common
ancestor of two or more descendant branches. All the descendant branches
of a given node (and all of the their respective descendants) are said
to form a <strong>clade</strong> (or <strong>monophyletic group</strong>).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># First we set a seed. This isn&#39;t specific to phylogenies, but means we use the </span>
<span class="c1"># same random numbers every time. This is because random numbers from your </span>
<span class="c1"># computer are generated using the internal clock. Google it for more info!</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">0</span><span class="p">)</span>

<span class="c1"># rtree creates a random tree.</span>
<span class="nf">plot</span><span class="p">(</span><span class="nf">rtree</span><span class="p">(</span><span class="m">10</span><span class="p">),</span> <span class="s">&quot;unrooted&quot;</span><span class="p">)</span>

<span class="c1"># Plot clade label.</span>
<span class="nf">rect</span><span class="p">(</span><span class="m">1.3</span><span class="p">,</span><span class="m">2.0</span><span class="p">,</span><span class="m">2.5</span><span class="p">,</span><span class="m">2.7</span><span class="p">,</span> <span class="n">border</span> <span class="o">=</span> <span class="s">&quot;grey&quot;</span><span class="p">)</span>
<span class="nf">text</span><span class="p">(</span><span class="m">2.8</span><span class="p">,</span> <span class="m">2.5</span><span class="p">,</span> <span class="s">&quot;Clade&quot;</span><span class="p">)</span>

<span class="c1"># Plot node label.</span>
<span class="nf">arrows</span><span class="p">(</span><span class="m">0.35</span><span class="p">,</span><span class="m">0.9</span><span class="p">,</span><span class="m">0.75</span><span class="p">,</span><span class="m">1</span><span class="p">,</span> <span class="n">length</span> <span class="o">=</span> <span class="m">0.125</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">20</span><span class="p">,</span> <span class="n">code</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
<span class="nf">text</span><span class="p">(</span><span class="m">1.0</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="s">&quot;Node&quot;</span><span class="p">)</span>

<span class="c1"># Plot edge label.</span>
<span class="nf">arrows</span><span class="p">(</span><span class="m">0.65</span><span class="p">,</span><span class="m">1.45</span><span class="p">,</span><span class="m">0.15</span><span class="p">,</span><span class="m">1.4</span><span class="p">,</span> <span class="n">length</span> <span class="o">=</span> <span class="m">0.125</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">20</span><span class="p">,</span> <span class="n">code</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
<span class="nf">text</span><span class="p">(</span><span class="m">-0.1</span><span class="p">,</span> <span class="m">1.4</span><span class="p">,</span> <span class="s">&quot;Edge&quot;</span><span class="p">)</span>

<span class="c1"># Plot tip label.</span>
<span class="nf">arrows</span><span class="p">(</span><span class="m">-0.15</span><span class="p">,</span><span class="m">2.8</span><span class="p">,</span><span class="m">-0.45</span><span class="p">,</span><span class="m">2.95</span><span class="p">,</span> <span class="n">length</span> <span class="o">=</span> <span class="m">0.125</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">20</span><span class="p">,</span> <span class="n">code</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
<span class="nf">text</span><span class="p">(</span><span class="m">-0.6</span><span class="p">,</span> <span class="m">3</span><span class="p">,</span> <span class="s">&quot;Tip&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-3-1.png"><img alt="../../_images/unnamed-chunk-3-1.png" class="align-center" src="../../_images/unnamed-chunk-3-1.png" style="width: 600px;" /></a>
<p>When we select a node to act as the base of a tree, the tree is said to
be <strong>rooted</strong>. At the bottom of a tree, is the <strong>root node</strong> (or simply
the <strong>root</strong>).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># You can also read in trees directly from text by specifying the branch </span>
<span class="c1"># lengths after each node/tip. More info below!</span>
<span class="n">tree</span> <span class="o">&lt;-</span> <span class="nf">read.tree</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="s">&quot;(((Homo:1, Pan:1):1, Gorilla:1):1, Pongo:1);&quot;</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>

<span class="c1"># Plot root label.</span>
<span class="nf">lines</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">-0.5</span><span class="p">,</span><span class="m">0</span><span class="p">),</span> <span class="nf">c</span><span class="p">(</span><span class="m">3.18</span><span class="p">,</span><span class="m">3.18</span><span class="p">))</span>
<span class="nf">arrows</span><span class="p">(</span><span class="m">0.03</span><span class="p">,</span><span class="m">3.18</span><span class="p">,</span><span class="m">0.35</span><span class="p">,</span><span class="m">3.18</span><span class="p">,</span> <span class="n">length</span> <span class="o">=</span> <span class="m">0.125</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">20</span><span class="p">,</span> <span class="n">code</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
<span class="nf">text</span><span class="p">(</span><span class="m">0.5</span><span class="p">,</span> <span class="m">3.18</span><span class="p">,</span> <span class="s">&quot;Root&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-4-1.png"><img alt="../../_images/unnamed-chunk-4-1.png" class="align-center" src="../../_images/unnamed-chunk-4-1.png" style="width: 600px;" /></a>
<p>Phylogenetic trees of the kind shown above are fairly simple and lack
information about time or character changes occurring along a branch. We
can assign branch length in the form of either time or the amount of
change/substitution along a branch. A tree with <strong>branch lengths</strong>
depicted can be called a <strong>phylogram</strong>.</p>
<p>When (an implied) dimension of time is being considered, all the tips of
the tree must be at the level representing the time in which they are
observed. For trees where all the species are extant, the tips are flush
at the top. This representation is called an <strong>ultrametric</strong> tree.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read in a text tree and plot.</span>
<span class="n">tree</span> <span class="o">&lt;-</span> <span class="nf">read.tree</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="s">&quot;(((Homo:6.3, Pan:6.3):2.5, Gorilla:8.8):6.9, Pongo:15.7);&quot;</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>

<span class="c1"># Add an axis along the bottom.</span>
<span class="nf">axisPhylo</span><span class="p">()</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-5-1.png"><img alt="../../_images/unnamed-chunk-5-1.png" class="align-center" src="../../_images/unnamed-chunk-5-1.png" style="width: 600px;" /></a>
</section>
<section id="informatic-representations-of-tree">
<h3>Informatic representations of tree<a class="headerlink" href="#informatic-representations-of-tree" title="Permalink to this headline">¶</a></h3>
<p>To perform any useful calculations on a tree, we need both a
computer-readable tree format and (in part) to understand how trees are
constructed in computer memory.</p>
</section>
<section id="text-based-formats">
<h3>Text based formats<a class="headerlink" href="#text-based-formats" title="Permalink to this headline">¶</a></h3>
<p>Storage of trees for transfer between different software is essential.
This is most commonly achieved with a text-based format stored in a
file. The most common file format for representing phylogenetic trees is
<strong>Newick format</strong>. This consists of clades represented within
parentheses. Commas separate each lade. Either tip names or symbols
representing the tips are nested within the lowest orders of
parentheses. Each tip or branch can be associated with a branch length
scalar that follows a colon.</p>
<p>For example:</p>
<p><code class="docutils literal notranslate"><span class="pre">&quot;(((Homo,</span> <span class="pre">Pan),</span> <span class="pre">Gorilla),</span> <span class="pre">Pongo);&quot;</span></code></p>
<p>Or with branch length:</p>
<p><code class="docutils literal notranslate"><span class="pre">&quot;(((Homo:6.3,</span> <span class="pre">Pan:6.3):2.5,</span> <span class="pre">Gorilla:8.8):6.9,</span> <span class="pre">Pongo:15.7);&quot;</span></code></p>
<p>Trees are also increasing use of XML formats such as PhyloXML and NeXML.</p>
<p>In this practical we are going to use the <code class="docutils literal notranslate"><span class="pre">elopomorph.tre</span></code> newick tree.
You can open it with a simple text editor to see the newick tree
structure.</p>
</section>
<section id="edge-table">
<h3>Edge table<a class="headerlink" href="#edge-table" title="Permalink to this headline">¶</a></h3>
<p>It is also possible to represent a phylogenetic tree as a matrix of
edges and vertices called an edge table. This is an even less intuitive
representation, but it is implemented in <code class="docutils literal notranslate"><span class="pre">R</span></code> and worth reviewing here.</p>
<p>There are a number of conventions that can be used to create an edge
table. The general concept consists of numbering the tips <em>1 - n</em>, and
all internal nodes labeled <em>n+1 … n+n-1</em>. The numbers for the internal
nodes can be assigned arbitrarily or according to an algorithm.</p>
<p>In <code class="docutils literal notranslate"><span class="pre">R</span></code> packages like <code class="docutils literal notranslate"><span class="pre">ape</span></code>, edge tables are constructed as follows:</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>node</p></th>
<th class="head"><p>connects to</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>5</p></td>
<td><p>6</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>7</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-odd"><td><p>7</p></td>
<td><p>2</p></td>
</tr>
<tr class="row-even"><td><p>6</p></td>
<td><p>3</p></td>
</tr>
<tr class="row-odd"><td><p>5</p></td>
<td><p>4</p></td>
</tr>
</tbody>
</table>
<p>You read the table as follows: node <code class="docutils literal notranslate"><span class="pre">5</span></code> (root) connects to node <code class="docutils literal notranslate"><span class="pre">6</span></code>. The
node <code class="docutils literal notranslate"><span class="pre">6</span></code> connects to node <code class="docutils literal notranslate"><span class="pre">7</span></code>. Node <code class="docutils literal notranslate"><span class="pre">7</span></code> connects to node <code class="docutils literal notranslate"><span class="pre">1</span></code> that happen
to be the first tip (<code class="docutils literal notranslate"><span class="pre">Homo</span></code>) and to node <code class="docutils literal notranslate"><span class="pre">2</span></code> (<code class="docutils literal notranslate"><span class="pre">Pan</span></code>) etc… Note that in a
binary tree (i.e. a tree where each node has only two descendants) each
node always connects to two elements (nodes or tips).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a tree.</span>
<span class="n">tree</span> <span class="o">&lt;-</span> <span class="nf">read.tree</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="s">&quot;(((Homo, Pan), Gorilla), Pongo);&quot;</span><span class="p">)</span>

<span class="c1"># Plot the tree.</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">label.offset</span> <span class="o">=</span> <span class="m">0.1</span><span class="p">)</span>

<span class="c1"># Add node labels and tip labels to the existing plot.</span>
<span class="nf">nodelabels</span><span class="p">()</span>
<span class="nf">tiplabels</span><span class="p">()</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-6-1.png"><img alt="../../_images/unnamed-chunk-6-1.png" class="align-center" src="../../_images/unnamed-chunk-6-1.png" style="width: 600px;" /></a>
</section>
<section id="records-pointers">
<h3>Records &amp; pointers<a class="headerlink" href="#records-pointers" title="Permalink to this headline">¶</a></h3>
<p>At a lower level, phylogenetic trees can be represented in computer
memory as more complex data objects. We don’t need to go into detail
here, but if you consider nodes and tips as data objects (i.e. a
dataframe), a tree could be stored as an array of dataframes which store
information about which members of that same array are descendants and
which are ancestors.</p>
</section>
</section>
<section id="basic-tree-viewing-in-r">
<h2>3. Basic tree viewing in <code class="docutils literal notranslate"><span class="pre">R</span></code><a class="headerlink" href="#basic-tree-viewing-in-r" title="Permalink to this headline">¶</a></h2>
<p>Now let’s visualise some phylogenies! We’ll use the Elopomorpha (eels
and similar fishes) tree to start as it is simple.</p>
<section id="reading-in-a-phylogeny-from-a-file">
<h3>Reading in a phylogeny from a file<a class="headerlink" href="#reading-in-a-phylogeny-from-a-file" title="Permalink to this headline">¶</a></h3>
<p>To load a tree you need the function <code class="docutils literal notranslate"><span class="pre">read.tree</span></code>. Just like we did with
text but instead we point to a file location. <code class="docutils literal notranslate"><span class="pre">read.tree</span></code> can read any
newick format trees (see above) like the <code class="docutils literal notranslate"><span class="pre">elopomorph.tre</span></code> file.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read in the tree from a file in the data folder.</span>
<span class="n">fishtree</span> <span class="o">&lt;-</span> <span class="nf">read.tree</span><span class="p">(</span><span class="s">&quot;data/elopomorph.tre&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Be sure you are always in the right directory. Remember you can navigate in <code class="docutils literal notranslate"><span class="pre">R</span></code>
using <code class="docutils literal notranslate"><span class="pre">setwd()</span></code>, <code class="docutils literal notranslate"><span class="pre">getwd()</span></code> and <code class="docutils literal notranslate"><span class="pre">list.files()</span></code> (to see what’s in the current directory).</p>
</div>
<p>Let’s examine the tree by typing:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Return the object.</span>
<span class="n">fishtree</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## 
## Phylogenetic tree with 62 tips and 61 internal nodes.
## 
## Tip labels:
##   Moringua_edwardsi, Kaupichthys_nuchalis, Gorgasia_taiwanensis, Heteroconger_hassi, Venefica_proboscidea, Anguilla_rostrata, ...
## 
## Rooted; includes branch lengths.
</pre></div>
</div>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get the structure of the object.</span>
<span class="nf">str</span><span class="p">(</span><span class="n">fishtree</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## List of 4
##  $ edge       : int [1:122, 1:2] 63 64 64 65 66 67 68 68 69 70 ...
##  $ edge.length: num [1:122] 0.0105 0.0672 0.00537 0.00789 0.00157 ...
##  $ Nnode      : int 61
##  $ tip.label  : chr [1:62] &quot;Moringua_edwardsi&quot; &quot;Kaupichthys_nuchalis&quot; &quot;Gorgasia_taiwanensis&quot; &quot;Heteroconger_hassi&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;phylo&quot;
##  - attr(*, &quot;order&quot;)= chr &quot;cladewise&quot;
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">fishtree</span></code> is a fully resolved tree with branch lengths. There are 62
species and 61 internal nodes. We can plot the tree by using the
<code class="docutils literal notranslate"><span class="pre">plot.phylo</span></code> function of <code class="docutils literal notranslate"><span class="pre">ape</span></code>. Note that we can just use the function
<code class="docutils literal notranslate"><span class="pre">plot</span></code> to do this as <code class="docutils literal notranslate"><span class="pre">R</span></code> knows if we ask it to plot a phylogeny to use
<code class="docutils literal notranslate"><span class="pre">plot.phylo</span></code> instead!</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the tree. </span>
<span class="nf">plot</span><span class="p">(</span><span class="n">fishtree</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">0.5</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-10-1.png"><img alt="../../_images/unnamed-chunk-10-1.png" class="align-center" src="../../_images/unnamed-chunk-10-1.png" style="width: 600px;" /></a>
<p><code class="docutils literal notranslate"><span class="pre">cex</span> <span class="pre">=</span> <span class="pre">0.5</span></code> reduces the size of the tip labels so we can read them.
(just about)</p>
<p>We can also zoom into different sections of the tree that you’re
interested in:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Zoom into the tree.</span>
<span class="nf">zoom</span><span class="p">(</span><span class="n">fishtree</span><span class="p">,</span> <span class="nf">grep</span><span class="p">(</span><span class="s">&quot;Gymnothorax&quot;</span><span class="p">,</span> <span class="n">fishtree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">),</span> <span class="n">subtree</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">0.8</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-11-1.png"><img alt="../../_images/unnamed-chunk-11-1.png" class="align-center" src="../../_images/unnamed-chunk-11-1.png" style="width: 600px;" /></a>
<p>The <code class="docutils literal notranslate"><span class="pre">grep</span></code> function is a generic function in <code class="docutils literal notranslate"><span class="pre">R</span></code> that allows to <em>grab</em>
any element in an object containing the desired characters. In this
example, <code class="docutils literal notranslate"><span class="pre">grep</span></code> is going to search for all the elements in
<code class="docutils literal notranslate"><span class="pre">fishtree$tip.label</span></code> that contains <code class="docutils literal notranslate"><span class="pre">Gymnothorax</span></code>
(e.g. <code class="docutils literal notranslate"><span class="pre">Gymnothorax_kidako</span></code>, <code class="docutils literal notranslate"><span class="pre">Gymnothorax_reticularis</span></code>). Try using only
<code class="docutils literal notranslate"><span class="pre">grep(&quot;thorax&quot;,</span> <span class="pre">fishtree$tip.label)</span></code> to see if it also only selects the
members of the <em>Gymnothorax</em> genus.</p>
<p>In this example, we just display the tree for the <em>Gymnothorax</em> genus
but you can also see how the species fit into the rest of the tree
using:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Pull out the Gymnothorax tips.</span>
<span class="nf">zoom</span><span class="p">(</span><span class="n">fishtree</span><span class="p">,</span> <span class="nf">grep</span><span class="p">(</span><span class="s">&quot;Gymnothorax&quot;</span><span class="p">,</span> <span class="n">fishtree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">),</span> <span class="n">subtree</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">0.8</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-12-1.png"><img alt="../../_images/unnamed-chunk-12-1.png" class="align-center" src="../../_images/unnamed-chunk-12-1.png" style="width: 600px;" /></a>
<p>Note that <code class="docutils literal notranslate"><span class="pre">zoom</span></code> is a specific plotting function that will automatically
set the plotting window to display two plots at once. This might create
some conflicts if you’re using RStudio. The bug can be easily solved
though by typing <code class="docutils literal notranslate"><span class="pre">dev.off()</span></code> to reinitialise the plotting window and
then proceed to the normal <code class="docutils literal notranslate"><span class="pre">zoom(...)</span></code> function as written above.</p>
<p>You can also reset this to one plot only per window by using:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Change the plotting parameters back to 1 row and 1 column.</span>
<span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">1</span><span class="p">))</span>
</pre></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>If you’re plots are coming out weird, resetting the plotting device is a good first
move. Use <code class="docutils literal notranslate"><span class="pre">dev.off()</span></code> first and see if it fixes it. A minor warning that it will
delete all the previous plots, but this is why we use scripts!</p>
</div>
<p>To get further options for the plotting of phylogenies:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Help function.</span>
<span class="o">?</span><span class="n">plot.phylo</span>
</pre></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Remeber that using the question mark (<code class="docutils literal notranslate"><span class="pre">?</span></code>) can also be done for every function.
The help pages may look confusing/intimidating at first but you quickly get used
to the set up.</p>
</div>
<p>Note that although you can use <code class="docutils literal notranslate"><span class="pre">plot</span></code> to plot the phylogeny, you need to
specify <code class="docutils literal notranslate"><span class="pre">plot.phylo</span></code> to find out the options for plotting trees. You can
change the style of the tree (<code class="docutils literal notranslate"><span class="pre">type</span></code>), the color of the branches and
tips (<code class="docutils literal notranslate"><span class="pre">edge.color</span></code>, <code class="docutils literal notranslate"><span class="pre">tip.color</span></code>), and the size of the tip labels
(<code class="docutils literal notranslate"><span class="pre">cex</span></code>). Here’s an fun/hideous example!</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot</span><span class="p">(</span><span class="n">fishtree</span><span class="p">,</span> <span class="n">type</span> <span class="o">=</span> <span class="s">&quot;unrooted&quot;</span><span class="p">,</span> <span class="n">edge.color</span> <span class="o">=</span> <span class="s">&quot;deeppink&quot;</span><span class="p">,</span> <span class="n">tip.color</span> <span class="o">=</span> <span class="s">&quot;springgreen&quot;</span><span class="p">,</span>  <span class="n">cex</span> <span class="o">=</span> <span class="m">0.7</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-17-1.png"><img alt="../../_images/unnamed-chunk-17-1.png" class="align-center" src="../../_images/unnamed-chunk-17-1.png" style="width: 600px;" /></a>
<p>Or try</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot</span><span class="p">(</span><span class="nf">ladderize</span><span class="p">(</span><span class="n">fishtree</span><span class="p">),</span> <span class="n">type</span> <span class="o">=</span> <span class="s">&quot;c&quot;</span><span class="p">,</span> <span class="n">edge.color</span> <span class="o">=</span> <span class="s">&quot;darkviolet&quot;</span><span class="p">,</span> <span class="n">tip.color</span> <span class="o">=</span> <span class="s">&quot;hotpink&quot;</span><span class="p">,</span>  <span class="n">cex</span> <span class="o">=</span> <span class="m">0.7</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-18-1.png"><img alt="../../_images/unnamed-chunk-18-1.png" class="align-center" src="../../_images/unnamed-chunk-18-1.png" style="width: 600px;" /></a>
<p>The <code class="docutils literal notranslate"><span class="pre">ladderize</span></code> function allows to display the branches from shortest to
longest.</p>
<blockquote>
<div><p>Try to modify the graphical options (colors, display, size, ordering
of the nodes, etc.) to obtain the most beautiful or ugliest
Elopomorpha phylogeny!</p>
</div></blockquote>
</section>
</section>
<section id="manipulating-phylogenetic-trees-in-r">
<h2>4. Manipulating phylogenetic trees in <code class="docutils literal notranslate"><span class="pre">R</span></code><a class="headerlink" href="#manipulating-phylogenetic-trees-in-r" title="Permalink to this headline">¶</a></h2>
<p>There are a range of ways in which we can manipulate trees in R. To
start lets take a look at the bird family Turdidae.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read in the turdidae (thrush) tree.</span>
<span class="n">turdidae_tree</span> <span class="o">&lt;-</span> <span class="nf">read.nexus</span><span class="p">(</span><span class="s">&quot;data/turdidae_birdtree.nex&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>As this multiphylo object (i.e. contains 100 different trees) we need to
first choose one random tree before we start.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Sample a random tree. Because we set our seed earlier this will be the same </span>
<span class="c1"># random tree for everyone. </span>
<span class="n">ran_turdidae_tree</span> <span class="o">&lt;-</span> <span class="nf">sample</span><span class="p">(</span><span class="n">turdidae_tree</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="m">1</span><span class="p">)[[</span><span class="m">1</span><span class="p">]]</span>

<span class="c1"># Use the plot tree function from phytools to plot a fan tree.</span>
<span class="nf">plotTree</span><span class="p">(</span><span class="n">ran_turdidae_tree</span><span class="p">,</span> <span class="n">type</span><span class="o">=</span><span class="s">&quot;fan&quot;</span><span class="p">,</span> <span class="n">fsize</span><span class="o">=</span><span class="m">0.4</span><span class="p">,</span> <span class="n">lwd</span><span class="o">=</span><span class="m">0.5</span><span class="p">,</span><span class="n">ftype</span><span class="o">=</span><span class="s">&quot;i&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-20-1.png"><img alt="../../_images/unnamed-chunk-20-1.png" class="align-center" src="../../_images/unnamed-chunk-20-1.png" style="width: 600px;" /></a>
<p>First, lets see what species are in the tree.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Lets see the first 10 species.</span>
<span class="nf">head</span><span class="p">(</span><span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">,</span> <span class="m">10</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>##  [1] &quot;Zoothera_everetti&quot;        &quot;Zoothera_naevia&quot;          &quot;Zoothera_pinicola&quot;       
##  [4] &quot;Hylocichla_mustelina&quot;     &quot;Catharus_aurantiirostris&quot; &quot;Catharus_mexicanus&quot;      
##  [7] &quot;Catharus_dryas&quot;           &quot;Catharus_fuscater&quot;        &quot;Catharus_ustulatus&quot;      
## [10] &quot;Catharus_bicknelli&quot;
</pre></div>
</div>
<p>We can remove species from the tree that we don’t want to include. This
is useful when we are missing data for a species, or we have a larger
tree with species we don’t need.</p>
<p>Lets say we want to drop all species with the Myadestes genus. In this
instance we first find all the associated tip.labels.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create an string object of the name we want to remove.</span>
<span class="n">drop_pattern</span> <span class="o">&lt;-</span> <span class="p">(</span><span class="s">&quot;Myadestes&quot;</span><span class="p">)</span>

<span class="c1"># sapply will iterate a given function over a vector (check out apply, lapply, </span>
<span class="c1"># mapply for more info)</span>
<span class="c1"># in this case, we&#39;re using the grep function to ask if any species name in our</span>
<span class="c1"># tip label list matches the drop.species pattern. </span>
<span class="n">tip_numbers</span> <span class="o">&lt;-</span> <span class="nf">sapply</span><span class="p">(</span><span class="n">drop_pattern</span><span class="p">,</span> <span class="n">grep</span><span class="p">,</span> <span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">)</span>

<span class="c1"># We then use the tip numbers to select only the tips we want.</span>
<span class="n">drop_species</span> <span class="o">&lt;-</span> <span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">[</span><span class="n">tip_numbers</span><span class="p">]</span>
<span class="n">drop_species</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>##  [1] &quot;Myadestes_genibarbis&quot;   &quot;Myadestes_ralloides&quot;    &quot;Myadestes_melanops&quot;     &quot;Myadestes_coloratus&quot;   
##  [5] &quot;Myadestes_palmeri&quot;      &quot;Myadestes_elisabeth&quot;    &quot;Myadestes_townsendi&quot;    &quot;Myadestes_obscurus&quot;    
##  [9] &quot;Myadestes_occidentalis&quot; &quot;Myadestes_unicolor&quot;     &quot;Myadestes_lanaiensis&quot;
</pre></div>
</div>
<p>Now, we create a new tree, with the Myadestes tips dropped from it.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># drop.tip will remove any matching tips from the tree.</span>
<span class="n">ran_turdidae_tree_NM</span> <span class="o">&lt;-</span> <span class="nf">drop.tip</span><span class="p">(</span><span class="n">ran_turdidae_tree</span><span class="p">,</span> <span class="n">drop_species</span><span class="p">)</span>

<span class="c1"># Plot the tree.</span>
<span class="nf">plotTree</span><span class="p">(</span><span class="n">ran_turdidae_tree_NM</span><span class="p">,</span> <span class="n">type</span><span class="o">=</span><span class="s">&quot;fan&quot;</span><span class="p">,</span> <span class="n">fsize</span><span class="o">=</span><span class="m">0.4</span><span class="p">,</span> <span class="n">lwd</span><span class="o">=</span><span class="m">0.5</span><span class="p">,</span> <span class="n">ftype</span><span class="o">=</span><span class="s">&quot;i&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-23-1.png"><img alt="../../_images/unnamed-chunk-23-1.png" class="align-center" src="../../_images/unnamed-chunk-23-1.png" style="width: 600px;" /></a>
<p>Alternatively, lets say we want to extract the clade within the tree
that includes the pre identified selected range of species. We’ll now
keep the drop species and remove everything else.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Set diff will find all the tips that don&#39;t match drop_species.</span>
<span class="n">species_we_dont_want</span> <span class="o">&lt;-</span> <span class="nf">setdiff</span><span class="p">(</span><span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">,</span> <span class="n">drop_species</span><span class="p">)</span>

<span class="c1"># Set diff will find all the tips that don&#39;t match drop_species, and then drop.tip will remove them</span>
<span class="n">pruned_birdtree</span> <span class="o">&lt;-</span> <span class="nf">drop.tip</span><span class="p">(</span><span class="n">ran_turdidae_tree</span><span class="p">,</span> <span class="n">species_we_dont_want</span><span class="p">)</span>

<span class="c1"># Plot the smaller tree.</span>
<span class="nf">plotTree</span><span class="p">(</span><span class="n">pruned_birdtree</span><span class="p">,</span> <span class="n">ftype</span><span class="o">=</span><span class="s">&quot;i&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-24-1.png"><img alt="../../_images/unnamed-chunk-24-1.png" class="align-center" src="../../_images/unnamed-chunk-24-1.png" style="width: 600px;" /></a>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>You can save your new revised tree. This can save time reading in a big tree just
to remove tips. You’ll notice in later practicals that the phylogenetic tree for
all birds is large, and takes a while to read into R, so it’s worth saving pruned trees!
You can save trees using <code class="docutils literal notranslate"><span class="pre">write.tree</span></code>, the same as you would save a dataframe.</p>
</div>
<p>For some analyses, you might want to work with on a genus level tree.
This can easily be done by a few key steps. We can do this using base
<code class="docutils literal notranslate"><span class="pre">R</span></code>, but we’ll use some packages from the <code class="docutils literal notranslate"><span class="pre">tidyverse</span></code> that have some
very useful functions. First load <code class="docutils literal notranslate"><span class="pre">stringr</span></code> (for manipulating strings)
and <code class="docutils literal notranslate"><span class="pre">dplyr</span></code> (for manipulating dataframes). You’ll see that it masks
functions from other packages like stats. If you need those functions
you can use code like this: <code class="docutils literal notranslate"><span class="pre">stats::filter()</span></code> to specify the package you
want.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load packages from the tidyverse. </span>
<span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Copy a list of all the tips from the tree.</span>
<span class="n">bird_tips</span> <span class="o">&lt;-</span> <span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span>

<span class="c1"># Split the labels into two strings where there&#39;s an underscore </span>
<span class="c1"># (simplfy returns the splits as separate columns in a dataframe)</span>
<span class="n">bird_genera</span> <span class="o">&lt;-</span> <span class="n">bird_tips</span> <span class="o">%&gt;%</span> <span class="nf">str_split</span><span class="p">(</span><span class="n">pattern</span> <span class="o">=</span> <span class="s">&quot;_&quot;</span><span class="p">,</span> <span class="n">simplify</span><span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">bird_genera</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Genus&quot;</span><span class="p">,</span> <span class="s">&quot;Species&quot;</span><span class="p">)</span>

<span class="c1"># Pull out the rows that have an distinct genus name. </span>
<span class="c1"># This will be the first instance in the dataframe for that genus.</span>
<span class="n">bird_genera</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="n">bird_genera</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">distinct</span><span class="p">(</span><span class="n">Genus</span><span class="p">,</span> <span class="n">.keep_all</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="n">bird_genera</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>##             Genus         Species
## 1        Zoothera        everetti
## 2      Hylocichla       mustelina
## 3        Catharus aurantiirostris
## 4    Entomodestes       coracinus
## 5      Cichlopsis      leucogenys
## 6          Turdus      mupinensis
## 7    Psophocichla    litsitsirupa
## 8      Nesocichla         eremita
## 9      Cataponera       turdoides
## 10         Cochoa        purpurea
## 11 Chlamydochaera        jefferyi
## 12       Geomalia       heinrichi
## 13         Sialia     currucoides
## 14      Myadestes      genibarbis
## 15   Neocossyphus         poensis
## 16     Stizorhina         fraseri
## 17         Alethe     fuelleborni
## 18      Myophonus      borneensis
## 19   Brachypteryx        stellata
## 20     Heinrichia       calligyna
</pre></div>
</div>
<p>We now have a dataframe of species, with one tip per each unique genus.</p>
<p>A few things to note about the ‘tidy’ code we just ran:</p>
<p><code class="docutils literal notranslate"><span class="pre">%&gt;%</span></code> is an operator used to ‘pipe’ an object into a function. Piping is
common in most computing languages so look it up for more information!
It can make code less clutered by separating the data from the function
you need to use. This isn’t unique to <code class="docutils literal notranslate"><span class="pre">tidyverse</span></code> functions, but you’ll
see it used a lot more in their documentation.</p>
<p>For the <code class="docutils literal notranslate"><span class="pre">distinct</span></code> function, we specified a column name without using
quotation marks. To make code easier to read, most tidy functions will
process column names (or other labels) this way.</p>
<p>We can now use our unique species from each genera to drop all the other
tips in the tree.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Combine the columns and add back in the underscore so they match the labels in the tree.</span>
<span class="n">genera_tips</span> <span class="o">&lt;-</span> <span class="nf">paste</span><span class="p">(</span><span class="n">bird_genera</span><span class="o">$</span><span class="n">Genus</span><span class="p">,</span> <span class="n">bird_genera</span><span class="o">$</span><span class="n">Species</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;_&quot;</span><span class="p">)</span>

<span class="c1"># Pull out the tips we want to drop.</span>
<span class="n">drop_tips</span> <span class="o">&lt;-</span> <span class="nf">setdiff</span><span class="p">(</span><span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">,</span> <span class="n">genera_tips</span><span class="p">)</span>

<span class="c1"># Remove all the species except one per genus.</span>
<span class="n">genera_tree</span> <span class="o">&lt;-</span> <span class="nf">drop.tip</span><span class="p">(</span><span class="n">ran_turdidae_tree</span><span class="p">,</span> <span class="n">drop_tips</span><span class="p">)</span>
<span class="nf">plotTree</span><span class="p">(</span><span class="n">genera_tree</span><span class="p">,</span> <span class="n">ftype</span><span class="o">=</span><span class="s">&quot;i&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-28-1.png"><img alt="../../_images/unnamed-chunk-28-1.png" class="align-center" src="../../_images/unnamed-chunk-28-1.png" style="width: 600px;" /></a>
<p>As the tree has dropped all but one species per genus, this means we
will finally need to also rename the tip labels as well to reflect this
change.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># It&#39;s definitely worth checking that the labels match up properly when you change tip labels.</span>
<span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">2</span><span class="p">))</span>
<span class="nf">plotTree</span><span class="p">(</span><span class="n">genera_tree</span><span class="p">,</span> <span class="n">ftype</span><span class="o">=</span><span class="s">&quot;i&quot;</span><span class="p">)</span>

<span class="c1"># Swap species names for genera.</span>
<span class="n">genera_tree</span><span class="o">$</span><span class="n">tip.label</span> <span class="o">&lt;-</span> <span class="n">bird_genera</span><span class="o">$</span><span class="n">Genus</span>
<span class="nf">plotTree</span><span class="p">(</span><span class="n">genera_tree</span><span class="p">,</span> <span class="n">ftype</span><span class="o">=</span><span class="s">&quot;i&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-29-1.png"><img alt="../../_images/unnamed-chunk-29-1.png" class="align-center" src="../../_images/unnamed-chunk-29-1.png" style="width: 600px;" /></a>
<p>It is important to note, that a big limitation with this approach is
that by selecting only one species per genus to keep, that you may run
the risk of unintentially dropping tips of species that are
paraphyletic. For example, Zoothera genus is spread throughout Turdidae
tree.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Zoom into zoothera tree.</span>
<span class="nf">plotTree</span><span class="p">(</span><span class="n">ran_turdidae_tree</span><span class="p">,</span> <span class="n">fsize</span><span class="o">=</span><span class="m">0.2</span><span class="p">,</span> <span class="n">lwd</span><span class="o">=</span><span class="m">0.2</span><span class="p">,</span> <span class="n">ftype</span><span class="o">=</span><span class="s">&quot;i&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-30-1.png"><img alt="../../_images/unnamed-chunk-30-1.png" class="align-center" src="../../_images/unnamed-chunk-30-1.png" style="width: 600px;" /></a>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">zoom</span><span class="p">(</span><span class="n">ran_turdidae_tree</span><span class="p">,</span> <span class="nf">grep</span><span class="p">(</span><span class="s">&quot;Zoothera&quot;</span><span class="p">,</span> <span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">),</span> <span class="n">subtree</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">0.8</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-30-2.png"><img alt="../../_images/unnamed-chunk-30-2.png" class="align-center" src="../../_images/unnamed-chunk-30-2.png" style="width: 600px;" /></a>
<p>This means that when collapsing a phylogenetic tree you run the risk of
miss representing the relationship between the different genera. The
only way to get round this is by:</p>
<ol class="simple">
<li><p>making sure check to see how paraphyletic your tree is at the start;
this can be done more easily by uploading and viewing your tree file
&#64; <a class="reference external" href="https://itol.embl.de/">https://itol.embl.de/</a>.</p></li>
<li><p>Renaming your conflicting paraphyletic clades within your phylogeny,
by altering the individual species names.</p></li>
</ol>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">[</span><span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="o">==</span><span class="s">&quot;Turdus_philomelos&quot;</span><span class="p">]</span><span class="o">&lt;-</span><span class="s">&quot;Turdus1_philomelos&quot;</span>
</pre></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>For your coursework you don’t need to worry about paraphyletic tips. It’s just
worth knowing that this can happen if you use trees in the future!</p>
</div>
</section>
<section id="adding-trait-data-to-trees-in-r">
<h2>5. Adding trait data to trees in <code class="docutils literal notranslate"><span class="pre">R</span></code><a class="headerlink" href="#adding-trait-data-to-trees-in-r" title="Permalink to this headline">¶</a></h2>
<p>Often basic tree plots in R are all you need for exploring data and your
analysis. However, for publications and presentations it may be useful
to plot trees with associated trait data. We will try plotting data with
a tree, using the package <code class="docutils literal notranslate"><span class="pre">ggtree</span></code>, and extension of <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code>.</p>
<p>For this exercise we will use the Turdidae tree (Thrushes) with some
data on different habitat types. Load in the data:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read in the data and check it worked.</span>
<span class="n">turdidae_data</span> <span class="o">&lt;-</span> <span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;data/turdidae_data.csv&quot;</span><span class="p">)</span>
<span class="nf">str</span><span class="p">(</span><span class="n">turdidae_data</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## &#39;data.frame&#39;:    173 obs. of  3 variables:
##  $ jetz_name: chr  &quot;Alethe choloensis&quot; &quot;Alethe diademata&quot; &quot;Alethe fuelleborni&quot; &quot;Alethe poliocephala&quot; ...
##  $ habitat  : chr  &quot;Dense&quot; &quot;Dense&quot; &quot;Dense&quot; &quot;Dense&quot; ...
##  $ body_mass: num  41.3 31.5 52 32.4 35.2 ...
</pre></div>
</div>
<p>We first need to match our data to the tip labels. First we need to put
an underscore in the names, and then match them to see if there’s any
name differences or missing species. In your coursework you will find
that occasionally because of taxonomic disagreements, you might be
missing species. The easiset way to solve this is by manually checking
tips that don’t match.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Replace the blank space with an underscore.</span>
<span class="n">turdidae_data</span><span class="o">$</span><span class="n">jetz_name</span> <span class="o">&lt;-</span> <span class="n">turdidae_data</span><span class="o">$</span><span class="n">jetz_name</span> <span class="o">%&gt;%</span> <span class="nf">str_replace</span><span class="p">(</span><span class="s">&quot; &quot;</span><span class="p">,</span> <span class="s">&quot;_&quot;</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">turdidae_data</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>##                 jetz_name   habitat body_mass
## 1       Alethe_choloensis     Dense     41.30
## 2        Alethe_diademata     Dense     31.54
## 3      Alethe_fuelleborni     Dense     52.00
## 4     Alethe_poliocephala     Dense     32.37
## 5       Alethe_poliophrys     Dense     35.20
## 6 Brachypteryx_hyperythra Semi-Open     78.00
</pre></div>
</div>
<p>We’ll use the <code class="docutils literal notranslate"><span class="pre">%in%</span></code> operator, which is useful checking if our species
are in the tip labels</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># %in% returns a list of logicals (TRUE or FALSE) for each object in the vector.</span>
<span class="n">turdidae_data</span><span class="o">$</span><span class="n">jetz_name</span> <span class="o">%in%</span> <span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>##   [1]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
##  [18]  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
##  [35]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
##  [52]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
##  [69]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
##  [86]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [103]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE
## [120]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [137]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [154]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [171]  TRUE  TRUE  TRUE
</pre></div>
</div>
<p>We can save the results, and use this to select the rows we need from
turdidae_data. We’ll use the <code class="docutils literal notranslate"><span class="pre">!</span></code> operator, which means NOT. So in this
case it’s the species that are not in the tip labels. This can also work
for lots of other functions as well so try experimenting!</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># See which species are NOT in the tip labels.</span>
<span class="n">index</span> <span class="o">&lt;-</span> <span class="o">!</span><span class="p">(</span><span class="n">turdidae_data</span><span class="o">$</span><span class="n">jetz_name</span> <span class="o">%in%</span> <span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">)</span>
<span class="n">turdidae_data</span><span class="p">[</span><span class="n">index</span><span class="p">,]</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>##              jetz_name habitat body_mass
## 23  Chaetops_aurantius    Open     41.10
## 24   Chaetops_frenatus    Open     45.60
## 117  Turdus_philomelos   Dense     67.74
</pre></div>
</div>
<p>One of them we renamed so we’ll change that back for our plots. The
others aren’t in our taxonomy. This is because they’ve been moved to
Chaetopidae, a new family of just rockjumpers. This often happens as
there are three main bird taxonomies: Jetz, Birdlife, and American
Ornithological Society. We’ll just remove them from our dataset.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Change the name back.</span>
<span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">[</span><span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="o">==</span><span class="s">&quot;Turdus1_philomelos&quot;</span><span class="p">]</span><span class="o">&lt;-</span><span class="s">&quot;Turdus_philomelos&quot;</span>

<span class="c1"># Get the species that ARE in the tips.</span>
<span class="n">index</span> <span class="o">&lt;-</span> <span class="n">turdidae_data</span><span class="o">$</span><span class="n">jetz_name</span> <span class="o">%in%</span> <span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span>

<span class="c1"># Select only these species.</span>
<span class="n">turdidae_data</span> <span class="o">&lt;-</span> <span class="n">turdidae_data</span><span class="p">[</span><span class="n">index</span><span class="p">,]</span>
</pre></div>
</div>
<p>Now we can drop the tips that don’t match and get plotting! Can you do
this yourself?</p>
<div class="dropdown admonition">
<p class="admonition-title">Show the answer…  </p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get the tips we don&#39;t want.</span>
<span class="n">drop_tips</span> <span class="o">&lt;-</span> <span class="nf">setdiff</span><span class="p">(</span><span class="n">ran_turdidae_tree</span><span class="o">$</span><span class="n">tip.label</span><span class="p">,</span> <span class="n">turdidae_data</span><span class="o">$</span><span class="n">jetz_name</span><span class="p">)</span>

<span class="c1"># Drop species.</span>
<span class="n">turdi_tree</span> <span class="o">&lt;-</span> <span class="nf">drop.tip</span><span class="p">(</span><span class="n">ran_turdidae_tree</span><span class="p">,</span> <span class="n">drop_tips</span><span class="p">)</span>
<span class="nf">plotTree</span><span class="p">(</span><span class="n">turdi_tree</span><span class="p">,</span> <span class="n">ftype</span><span class="o">=</span><span class="s">&quot;i&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-38-1.png"><img alt="../../_images/unnamed-chunk-38-1.png" class="align-center" src="../../_images/unnamed-chunk-38-1.png" style="width: 600px;" /></a>
</div>
<p>Now lets try using ggtree to plot our data with our phylogeny.</p>
<p>If you’re using your own laptop, you need to install <code class="docutils literal notranslate"><span class="pre">ggtree</span></code>. Because
not all packages are available from <code class="docutils literal notranslate"><span class="pre">CRAN</span></code> directly through <code class="docutils literal notranslate"><span class="pre">R</span></code>, we’ll
install <code class="docutils literal notranslate"><span class="pre">BiocManager</span></code>. This is a package manager that can install
packages from other servers and is used a lot in bioinformatics.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;ggtree&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Notice that we used <code class="docutils literal notranslate"><span class="pre">::</span></code> to specify we want to use the <code class="docutils literal notranslate"><span class="pre">BiocManager</span></code> package to
install <code class="docutils literal notranslate"><span class="pre">ggtree</span></code>. Using <code class="docutils literal notranslate"><span class="pre">::</span></code> means we don’t have to load the package, and can be
handy when we only need to use a function once.</p>
</div>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load plotting packages.</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggtree</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">ggtree</span></code> is a bit more complicated than just normal tree plots, but you
can also do a lot more. We’ll create a basic tree plot structure first
and then add tip labels and traits after.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a plot of the tree with a circular layout.</span>
<span class="n">turdidae_plot</span> <span class="o">&lt;-</span> <span class="nf">ggtree</span><span class="p">(</span><span class="n">turdi_tree</span><span class="p">,</span> <span class="n">layout</span> <span class="o">=</span> <span class="s">&quot;circular&quot;</span><span class="p">)</span>

<span class="c1"># Return the plot. This will show our plot.</span>
<span class="n">turdidae_plot</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-42-11.png"><img alt="../../_images/unnamed-chunk-42-11.png" class="align-center" src="../../_images/unnamed-chunk-42-11.png" style="width: 600px;" /></a>
<p>This is a pretty basic plot. We can add tip labels and customize our
plot the same way as if we were using <code class="docutils literal notranslate"><span class="pre">ggplot2</span></code>.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a plot of the tree with a circular layout.</span>
<span class="n">turdidae_plot</span> <span class="o">&lt;-</span> <span class="nf">ggtree</span><span class="p">(</span><span class="n">turdi_tree</span><span class="p">,</span> <span class="n">layout</span> <span class="o">=</span> <span class="s">&quot;circular&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">geom_tiplab</span><span class="p">()</span>

<span class="c1"># Return the plot. This will show our plot.</span>
<span class="n">turdidae_plot</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-43-11.png"><img alt="../../_images/unnamed-chunk-43-11.png" class="align-center" src="../../_images/unnamed-chunk-43-11.png" style="width: 600px;" /></a>
<p>Our plot is now a bit big and falling out of the plotting window. We can
fix this by expanding the x axis, which unlike normal plots comes out
from the middle. We can also make the tips smaller as well.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a plot of the tree with a circular layout.</span>
<span class="n">turdidae_plot</span> <span class="o">&lt;-</span> <span class="nf">ggtree</span><span class="p">(</span><span class="n">turdi_tree</span><span class="p">,</span> <span class="n">layout</span> <span class="o">=</span> <span class="s">&quot;circular&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">geom_tiplab</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">1.5</span><span class="p">)</span>

<span class="c1"># Return the plot. This will show our plot.</span>
<span class="n">turdidae_plot</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-44-1.png"><img alt="../../_images/unnamed-chunk-44-1.png" class="align-center" src="../../_images/unnamed-chunk-44-1.png" style="width: 600px;" /></a>
<p>If we want to add trait data to our plot, we can combine our tree and
data together.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add a column that&#39;s the node number matching the tree.</span>
<span class="n">turdidae_data</span><span class="o">$</span><span class="n">node</span> <span class="o">&lt;-</span> <span class="nf">nodeid</span><span class="p">(</span><span class="n">turdi_tree</span><span class="p">,</span> <span class="n">turdidae_data</span><span class="o">$</span><span class="n">jetz_name</span><span class="p">)</span>

<span class="c1"># Join our data and tree together.</span>
<span class="n">turdi_tree_data</span> <span class="o">&lt;-</span>  <span class="nf">full_join</span><span class="p">(</span><span class="n">turdi_tree</span><span class="p">,</span> <span class="n">turdidae_data</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s">&quot;node&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Now we can make our plot!</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="n">turdidae_plot</span> <span class="o">&lt;-</span> <span class="nf">ggtree</span><span class="p">(</span><span class="n">turdi_tree_data</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s">&quot;fan&quot;</span><span class="p">,</span> <span class="n">open.angle</span> <span class="o">=</span> <span class="m">15</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">colour</span><span class="o">=</span><span class="n">habitat</span><span class="p">))</span> <span class="o">+</span> 
   <span class="nf">geom_tiplab</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">1.5</span><span class="p">)</span> <span class="o">+</span>
   <span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;firebrick&quot;</span><span class="p">,</span> <span class="s">&quot;navy&quot;</span><span class="p">,</span> <span class="s">&quot;forest green&quot;</span><span class="p">),</span> <span class="n">breaks</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Dense&quot;</span><span class="p">,</span> <span class="s">&quot;Semi-Open&quot;</span><span class="p">,</span> <span class="s">&quot;Open&quot;</span><span class="p">)))</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## Scale for &#39;y&#39; is already present. Adding another scale for &#39;y&#39;, which will replace the existing scale.
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-46-11.png"><img alt="../../_images/unnamed-chunk-46-11.png" class="align-center" src="../../_images/unnamed-chunk-46-11.png" style="width: 600px;" /></a>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Notice when we put an extra <code class="docutils literal notranslate"><span class="pre">()</span></code> around the entire line of code, it saves the plot
as an object, and prints the plot at the same time.</p>
</div>
<p>And now we have a plot where we can see the spread of habitat types in
Thrushes. Try experimenting with different colours and sizes to create
some beautiful trees that put this one to shame! There’s also lots of
other ways you can label trees. For more info this guide is a great
place to start:
<a class="reference external" href="https://4va.github.io/biodatasci/r-ggtree.html#the_ggtree_package">https://4va.github.io/biodatasci/r-ggtree.html#the_ggtree_package</a></p>
<blockquote>
<div><p>Extra task: Can you make another plot for continuous body mass data?
Think about how you’d show this with colour. Can you change the key to
better display the change in body mass? Is body mass more clumped in
the tree than habitat?</p>
</div></blockquote>
<div class="dropdown admonition">
<p class="admonition-title">Show the answer…  </p>
<p>We can try using points on the end of tips.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># There&#39;s lots of code here for the theme, which puts the legend in the middle.</span>
<span class="p">(</span><span class="n">turdidae_plot</span> <span class="o">&lt;-</span> <span class="nf">ggtree</span><span class="p">(</span><span class="n">turdi_tree_data</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s">&quot;fan&quot;</span><span class="p">,</span> <span class="n">open.angle</span> <span class="o">=</span> <span class="m">15</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="m">0.7</span><span class="p">)</span> <span class="o">+</span> 
   
   <span class="c1"># Make sure the tip labels are coloured black, and offset from the tree.</span>
   <span class="nf">geom_tiplab</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">1.5</span><span class="p">,</span> <span class="n">colour</span> <span class="o">=</span> <span class="s">&quot;black&quot;</span><span class="p">,</span> <span class="n">offset</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span> <span class="o">+</span> 
   
   <span class="c1"># Add the tip points.</span>
   <span class="nf">geom_tippoint</span><span class="p">(</span><span class="n">mapping</span><span class="o">=</span><span class="nf">aes</span><span class="p">(</span><span class="n">colour</span><span class="o">=</span><span class="n">body_mass</span><span class="p">),</span> <span class="n">size</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">show.legend</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span> <span class="o">+</span> 
   
   <span class="c1"># This is all for legend placement and size.</span>
   <span class="nf">theme</span><span class="p">(</span><span class="n">legend.title</span><span class="o">=</span><span class="nf">element_blank</span><span class="p">(),</span> 
        <span class="n">legend.position</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">0.65</span><span class="p">,</span><span class="m">0.45</span><span class="p">),</span> <span class="n">legend.direction</span> <span class="o">=</span> <span class="s">&quot;horizontal&quot;</span><span class="p">,</span> <span class="n">legend.title.align</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span>
        <span class="n">legend.key.width</span> <span class="o">=</span> <span class="nf">unit</span><span class="p">(</span><span class="m">1.2</span><span class="p">,</span> <span class="s">&quot;cm&quot;</span><span class="p">),</span> <span class="n">legend.key.height</span> <span class="o">=</span> <span class="nf">unit</span><span class="p">(</span><span class="m">0.16</span><span class="p">,</span> <span class="s">&quot;cm&quot;</span><span class="p">),</span> 
        <span class="n">legend.text</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">14</span><span class="p">),</span> <span class="n">legend.margin</span> <span class="o">=</span> <span class="kc">NULL</span><span class="p">)</span> <span class="o">+</span>
   
   <span class="c1"># Add the colour scheme.</span>
   <span class="nf">scale_colour_viridis_c</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s">&quot;Body Mass&quot;</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## Scale for &#39;y&#39; is already present. Adding another scale for &#39;y&#39;, which will replace the existing scale.
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-48-1.png"><img alt="../../_images/unnamed-chunk-48-1.png" class="align-center" src="../../_images/unnamed-chunk-48-1.png" style="width: 600px;" /></a>
<p>Notice how we use <code class="docutils literal notranslate"><span class="pre">open.angle</span> <span class="pre">=</span> <span class="pre">15</span></code> so that we can fit the legend in the
gap. Think how you can play around with the size of each part of the
plot to make it look nicer!</p>
</div>
<div class="dropdown admonition">
<p class="admonition-title">Show the answer…  </p>
<p>Or we can use bars. Maybe we could add clade labels after?</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use the ggtree extra package for adding plots to circular trees.</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggtreeExtra</span><span class="p">)</span>
<span class="p">(</span><span class="n">turdidae_plot</span> <span class="o">&lt;-</span> <span class="nf">ggtree</span><span class="p">(</span><span class="n">turdi_tree_data</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s">&quot;fan&quot;</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="m">0.7</span><span class="p">)</span> <span class="o">+</span>
    
    <span class="c1"># Geom fruit allows us to specify the ggplot geom we want.</span>
    <span class="nf">geom_fruit</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">turdi_tree_data</span><span class="p">,</span> <span class="n">geom</span><span class="o">=</span><span class="n">geom_bar</span><span class="p">,</span>
               <span class="n">mapping</span><span class="o">=</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">node</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="n">body_mass</span><span class="p">),</span>
               <span class="n">pwidth</span><span class="o">=</span><span class="m">0.38</span><span class="p">,</span>
               <span class="n">orientation</span><span class="o">=</span><span class="s">&quot;y&quot;</span><span class="p">,</span> 
               <span class="n">stat</span><span class="o">=</span><span class="s">&quot;identity&quot;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="s">&quot;navy&quot;</span><span class="p">,</span> <span class="n">colour</span><span class="o">=</span><span class="s">&quot;navy&quot;</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="m">0.2</span><span class="p">)</span> <span class="o">+</span> 
    <span class="c1"># We can remove some of the white space around our plot by setting the margins to negative values.</span>
  <span class="nf">theme</span><span class="p">(</span><span class="n">plot.margin</span><span class="o">=</span><span class="nf">margin</span><span class="p">(</span><span class="m">-80</span><span class="p">,</span><span class="m">-80</span><span class="p">,</span><span class="m">-80</span><span class="p">,</span><span class="m">-80</span><span class="p">)))</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## Scale for &#39;y&#39; is already present. Adding another scale for &#39;y&#39;, which will replace the existing scale.
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/unnamed-chunk-49-1.png"><img alt="../../_images/unnamed-chunk-49-1.png" class="align-center" src="../../_images/unnamed-chunk-49-1.png" style="width: 600px;" /></a>
<p>Geom fruit can be confusing as we have the y value as the node, and x as
our variable, but this code can easily be subbed in with your own data.</p>
<p>This guide goes into lots of detail plotting with ggtree, and is worth
your time! <a class="reference external" href="https://yulab-smu.top/treedata-book/chapter10.html">https://yulab-smu.top/treedata-book/chapter10.html</a></p>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./markdowns/practical_2"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../practical_1/practical_1.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Basics in R</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../practical_3/practical_3.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Macroecology analyses</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The Jupyter Book community<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>